[ID Refactor] Rename DriverID to JobID

jovany-wang opened the pull request:
## What do these changes do?

According to the [doc]( https://docs.google.com/document/d/1eXbwzRNgnmjiVeRlOzdKHUUquTxro8fI1PTMdKwMeGc/edit
), we'd like to change the `Driver` concept to `Job` to represent a job submitted by users. 
And the `Driver` concept is to represent a program to be performed of the job. We can treat driver as a special worker.

This PR is intended to:
1. On the `Job` concept,  rename `DriverID` to `JobID` on the `Job` concept.
2. On the `Driver ` concept, it's a special worker, so the type of a driver_id should be `WorkerID` naturally.

#### TODOs:
- [x] cpp
- [x] Java
- [x] Python

## Related issue number


## Linter

- [x] I've run `scripts/format.sh` to lint the changes in this PR.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-Perf-Integration-PRB/1291/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/14779/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/14781/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-Perf-Integration-PRB/1293/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-Perf-Integration-PRB/1294/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/14782/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-Perf-Integration-PRB/1344/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/14832/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-Perf-Integration-PRB/1348/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/14836/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/14852/
Test FAILed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/14854/
Test PASSed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/14859/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/14860/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/14858/
Test FAILed.

jovany-wang commented:
@zhijunfu

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-Perf-Integration-PRB/1364/
Test FAILed.

stephanie-wang commented on src/ray/core_worker/context.cc:
```diff
@@ -38,12 +38,11 @@ struct WorkerThreadContext {
-                    ? ClientID::FromBinary(driver_id.Binary())
-                    : ClientID::FromRandom()),
-      current_driver_id(worker_type == WorkerType::DRIVER ? driver_id : DriverID::Nil()) {
+      worker_id(worker_type == WorkerType::DRIVER ? ClientID::FromBinary(job_id.Binary())
```
I think these should be `WorkerID`, not `ClientID`.

stephanie-wang commented on src/ray/core_worker/core_worker.cc:
```diff
@@ -6,15 +6,15 @@ namespace ray {
                        const enum WorkerLanguage language,
                        const std::string &store_socket, const std::string &raylet_socket,
-                       DriverID driver_id)
+                       JobID job_id)
```
```suggestion
                       const JobID &job_id)
```

stephanie-wang commented on src/ray/raylet/scheduling_queue.cc:
```diff
@@ -187,9 +186,9 @@ void SchedulingQueue::FilterState(std::unordered_set<TaskID> &task_ids,
   } break;
   case TaskState::DRIVER: {
-    const auto driver_ids = GetDriverTaskIds();
+    const auto job_ids = GetDriverTaskIds();
```
Maybe rename to `driver_task_ids`?

stephanie-wang commented on src/ray/raylet/scheduling_queue.cc:
```diff
@@ -394,15 +392,15 @@ void SchedulingQueue::RemoveBlockedTaskId(const TaskID &task_id) {
-void SchedulingQueue::AddDriverTaskId(const TaskID &driver_id) {
-  RAY_LOG(DEBUG) << "Added driver task " << driver_id;
-  auto inserted = driver_task_ids_.insert(driver_id);
+void SchedulingQueue::AddDriverTaskId(const TaskID &job_id) {
```
Rename to `driver_task_id` (same below).

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-Perf-Integration-PRB/1365/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/14881/
Test FAILed.

jovany-wang commented on java/runtime/src/main/java/org/ray/runtime/WorkerContext.java:
```diff
@@ -52,13 +52,15 @@ public WorkerContext(WorkerMode workerMode, UniqueId driverId, RunMode runMode)
-      workerId = driverId;
+      // TODO(qwang): Assign the driver id to worker id
+      // once we treat driver id as a special worker id.
+      workerId = jobId;
```
I don't think we should set it to random id now.

In the one of the next several PRs, we should embedding job id to worker id if the mode is DRIVER. Because we should use the variable `workerId` represent the driver id or the worker id in different mode.

raulchen commented on python/ray/function_manager.py:
```diff
@@ -434,15 +434,15 @@ def f():
                     "function ID {}. Traceback:\n{}".format(
                         function_name, function_id.hex(), traceback_str),
-                    driver_id=driver_id)
+                    driver_id=ray.utils.compute_driver_id_from_job_id(job_id))
```
yeah, we should change the keyword arg as well

raulchen commented on python/ray/state.py:
```diff
@@ -1095,7 +1095,8 @@ def errors(include_cluster_errors=True):
+        driver_id=ray.utils.compute_driver_id_from_job_id(worker.current_job_id))
     if include_cluster_errors:
-        error_messages += state.error_messages(driver_id=ray.DriverID.nil())
+        error_messages += state.error_messages(driver_id=ray.WorkerID.nil())
```
we change change `driver_id` in `ErrorTable` to `job_id` as well.

raulchen commented on python/ray/tests/test_basic.py:
```diff
@@ -2614,8 +2614,8 @@ def f():
 def test_specific_driver_id():
-    dummy_driver_id = ray.DriverID(b"00112233445566778899")
-    ray.init(num_cpus=1, driver_id=dummy_driver_id)
+    dummy_driver_id = ray.JobID(b"00112233445566778899")
```
change the test name as well?

raulchen commented on python/ray/worker.py:
```diff
@@ -747,7 +748,7 @@ def run_function_on_all_workers(self, function,
             self.redis_client.hmset(
                 key, {
-                    "driver_id": self.task_driver_id.binary(),
+                    "job_id": ray.utils.compute_driver_id_from_job_id(self.current_job_id).binary(),
```
why `compute_driver_id_from_job_id`?

raulchen commented on python/ray/worker.py:
```diff
@@ -1781,6 +1782,7 @@ def connect(node,
         import __main__ as main
         driver_info = {
             "node_ip_address": node.node_ip_address,
+            # TODO(qwang): driver_id
```
explain more about this TODO

raulchen commented on src/ray/gcs/tables.cc:
```diff
@@ -450,37 +449,37 @@ Status Hash<ID, Data>::Subscribe(const DriverID &driver_id, const ClientID &clie
 }
 
-Status ErrorTable::PushErrorToDriver(const DriverID &driver_id, const std::string &type,
+Status ErrorTable::PushErrorToDriver(const WorkerID &driver_id, const std::string &type,
```
This argument can be a 'JobID', which means that we push errors to the driver of the given job.

raulchen commented on src/ray/raylet/format/node_manager.fbs:
```diff
@@ -135,6 +135,7 @@ table RegisterClientRequest {
   // The process ID of this worker.
   worker_pid: long;
   // The driver ID. This is non-nil if the client is a driver.
+  // TODO(qwang): rename this to driver_task_id.
```
This can also be `job_id`?

raulchen commented on src/ray/raylet/node_manager.h:
```diff
@@ -326,12 +326,12 @@ class NodeManager : public rpc::NodeManagerServiceHandler {
       const ray::gcs::ActorTable::WriteCallback &failure_callback);
 
-  /// When a driver dies, loop over all of the queued tasks for that driver and
+  /// When a job exited(or finished), loop over all of the queued tasks for that job and
```
```suggestion
  /// When a job finished, loop over all of the queued tasks for that job and
```

raulchen commented on src/ray/raylet/node_manager.h:
```diff
@@ -326,12 +326,12 @@ class NodeManager : public rpc::NodeManagerServiceHandler {
+  /// \param job_id The job that exited.
   /// \return Void.
-  void CleanUpTasksForDeadDriver(const DriverID &driver_id);
+  void CleanUpTasksForExitedJob(const JobID &job_id);
```
```suggestion
  void CleanUpTasksForFinishedJob(const JobID &job_id);
```

raulchen commented on src/ray/raylet/node_manager.cc:
```diff
@@ -210,18 +210,19 @@ ray::Status NodeManager::RegisterGcs() {
-      [this](gcs::AsyncGcsClient *client, const DriverID &client_id,
-             const std::vector<DriverTableData> &driver_data) {
-        HandleDriverTableUpdate(client_id, driver_data);
+  // TODO(qwang): rename client id parameter.
```
what does this todo mean?

jovany-wang commented on java/runtime/src/main/java/org/ray/runtime/raylet/RayletClientImpl.java:
```diff
@@ -44,10 +44,12 @@
   private long client = 0;
 
+  // TODO(qwang): Remove the parameter `jobId`
+  // once we treat driver id as a special worker id.
```
Yes!

raulchen replied:
Well, I think it's okay to still pass in the jobId, to make it more explicit. 
But if you do want to remove it, maybe refine the comment? E.g., `jobId` parameter can be removed once we embed `jobId` in `driverId`.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/14886/
Test PASSed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/14889/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-Perf-Integration-PRB/1370/
Test FAILed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/14892/
Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/14894/
Test PASSed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-Perf-Integration-PRB/1371/
Test FAILed.

jovany-wang commented:
Thanks all.
 I think I addressed all comments.

Btw, I think we should review this PR as quickly as possible to avoid more conflicts for other developers.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/14899/
Test PASSed.

stephanie-wang commented on python/ray/utils.py:
```diff
@@ -220,6 +221,11 @@ def binary_to_task_id(binary_task_id):
     return ray.TaskID(binary_task_id)
 
 
+def compute_driver_id_from_job_id(job_id):
```
I think we can remove this function now.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-Perf-Integration-PRB/1372/
Test FAILed.

raulchen commented on src/ray/common/id.h:
```diff
@@ -152,13 +152,19 @@ std::ostream &operator<<(std::ostream &os, const ObjectID &id);
+const WorkerID ComputeDriverId(const JobID &job_id);
+
+/// A helper function to compute a job id from the given driver id.
+const JobID ComputeJobId(const WorkerID &driver_id);
```
These 2 functions are no longer needed as well?

jovany-wang replied:
Nice catch!

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/14909/
Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/14910/
Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/14913/
Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/14914/
Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/14920/
Test PASSed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/14932/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-Perf-Integration-PRB/1378/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-Perf-Integration-PRB/1379/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-Perf-Integration-PRB/1382/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-Perf-Integration-PRB/1383/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-Perf-Integration-PRB/1389/
Test FAILed.
